HoP Pocketwatch

Flaborized opened the pull request:
<!-- The text between the arrows are comments - they won't be visible on your PR. -->
<!-- To label this PR, add the label(s) without the prefixes surrounded by brackets anywhere, for example: [LABEL] -->
<!-- PRs should at least have one area (A-) label and at least one category (C-) label -->

[FEAT]

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

This PR adds a pocketwatch called the "T-MO Watch" for the HoP, which can measure time and various financial stuff currently including:

- Current Overall Station Budget
- QM, Research, Payroll Budgets
- Total PTL earnings this round
- A step counter (counts your tiles moved while in pockets or hands, thanks gerhazo and pali for big code help)
- A timer (counts up in seconds, code borrowed heavily from game clocks)
- A light on the side of the watch that flickers when the HoP service bell is rung
- An HoP service bell ring counter
- Two silly nonsense modes unlocked with an emag
- If you smack someone with it you wave it in front of their face like a silly person

This is displayed using maptext over the item when held in-hand, and the watch even flicks open when you put it in your hand (and flicks shut when you drop it or put it in your pocket or w/e). The timekeeping uses the in-game timezone (same one clockwork oranges use). The display looks like an analog watch when it's measuring time, and a digital display when measuring other things (it's not a real analog watch, just a cool screen). Pictures:


![](https://cdn.discordapp.com/attachments/1041726307584704532/1106631706733191330/image.png)


![](https://cdn.discordapp.com/attachments/1041726307584704532/1106624800769712258/image.png)


![](https://cdn.discordapp.com/attachments/1041726307584704532/1106624875105357854/image.png)


![](https://cdn.discordapp.com/attachments/1041726307584704532/1106632015027126272/hop_watch_animate.gif)


![](https://cdn.discordapp.com/attachments/1041726307584704532/1106632068508680222/hop_watch_idle.gif)


![](https://cdn.discordapp.com/attachments/1041726307584704532/1107348836189487155/image.png)

(pretend all the pictures have the little light on the corner, thank you)

## Why's this needed? <!-- Describe why you think this should be added to the game. -->

I thought of the idea and thought it was cute and cool! The HoP could use a fun little item, and I think this one is in line with the design of the job, focused on money and probably the type to carry a fancy pocket watch. I also personally like being able to monitor how different departments are doing in a round, and this lets you do that in style while on the go. 

Also I think it's just an on-brand item for the HoP to have in their possession. It's like those expensive watches with tons of features, but it's also a pocketwatch which is just silly levels of weath-flaunting. Lets them track all kinds of little stats too, and lets them know when someone is ringing their bell.

## Changelog <!-- If necessary, put your changelog entry below. Otherwise, /please/ delete this entire section. -->
<!-- Put how you want to be credited in the changelog in place of CodeDude. -->
<!-- Use (*) for major changes and (+) for minor changes. See the contributor guide for details. For example: -->

```changelog
(u)Flaborized
(+)The HoP now carries a "T-MO Watch" on their person: it can measure a lot more than time, with many modes that can be swapped through by using it in-hand. Thanks to grifflez for helping with the name!
```

Piesuu commented:
It could take the amount of manifest entries (amount of security records) and display the amount of crew on the station from that!

Gannets commented:
Yeah, this feels on-brand for the HoP.

Flaborized commented:
some more work:

- added a step counter! it counts your steps (tiles moved by any means) while in your hand or pocket. made possible with good coding advice from pali and gerhazo
- named it the "T-MO Watch" based on a pun by grifflez. Time is money!!
- Added a 'timer' function where it counts up seconds.
- added a little light on the watch that blinks whenever someone rings the HoP service bell (no noise or chat output)
- generally rewrote some code that I hastily made to comply with the guidelines a bit more (though it's probably not perfect by any means)

Flaborized commented:
Undrafting this because I think I'm done adding stuff to it. Didn't make a stopwatch function since I didn't wanna do that Work, but may do it later or smile as someone else does it later. Same with crew headcount, good idea for future mayb

TobleroneSwordfish commented on code/obj/item/device/pocketwatch.dm:
```diff
@@ -0,0 +1,182 @@
+	object_flags = NO_GHOSTCRITTER
+	inventory_counter_enabled = TRUE
+	var/current_clock_mode = "Time Keeping"
+	var/hour_minute_divider = ":" //used to make the colon in the time blink like a clock display
```
these should be doc comments like
```
/// Comment
var/foo
```
or
```
var/foo //! Comment
```

TobleroneSwordfish commented on code/obj/item/device/pocketwatch.dm:
```diff
@@ -0,0 +1,182 @@
+			src.alert_overlay = image(src.icon)
+			src.alert_overlay.appearance_flags = PIXEL_SCALE | RESET_COLOR | RESET_ALPHA
+		src.alert_overlay.icon_state = "watch_alert_on"
+		src.UpdateOverlays(src.alert_overlay, "alert")
```
```suggestion
		src.UpdateOverlays(src.alert_overlay, "alert")
		if(current_clock_mode == "Service Bell Ring Counter")
			update_clock()
```

TobleroneSwordfish commented on code/obj/item/bell.dm:
```diff
@@ -17,6 +18,9 @@
 	src.visible_message("<span class='notice'><b>[user]</b> rings \the [src]!</span>")
 	playsound(src, 'sound/effects/bell_ring.ogg', 30, 0)
+	if(sends_signal_to_hop_watch)
+		for(var/obj/item/pocketwatch/watch in world)
```
iterating world is Bad, use `for_by_tcl` instead (you'll need to add `START_TRACKING` to the watch's `New` and `STOP_TRACKING` to its `disposing`)

TobleroneSwordfish commented on code/obj/item/device/pocketwatch.dm:
```diff
@@ -0,0 +1,182 @@
+			start_counting_steps(loc)
+		..()
+
+	proc/format_time_text(var/timeValue as num)
+		var/seconds = round((timeValue / 10) % 60)
+		var/minutes = round(((timeValue / 10) - seconds) / 60)
+		if (minutes < 10)
+			minutes = "0[minutes]"
+		if (seconds < 10)
+			seconds = "0[seconds]"
+		return "[minutes]:[seconds]"
```
since this is stolen from game clocks it should probably just go in `helpers.dm`

TobleroneSwordfish commented on code/obj/item/device/pocketwatch.dm:
```diff
@@ -0,0 +1,182 @@
+
+			if("Jones Quantity") //joke emag-only mode
+				var/amount_of_jones
+				for(var/mob/living/critter/small_animal/cat/jones/jones in world)
```
same here, `in world` is bad

TobleroneSwordfish reviewed (changes requested):
I love this thing, the little animations are great.

Flaborized commented:
> These are good sensible changes, though one issue I found testing them is that the time keeping function shows the wrong time entirely :sheltersweat: which I didn't notice cuz I originally implemented it with world.timeofday and broke it when switching to TIME so I gotta fix that

Ok fixed this, I hadn't realized that world.timeofday was what I needed for getting real time, and the guidelines made me think I should replace those uses with TIME. Thanks to pali and zewaka for helping figure this out!
